ykcompile/src/lib.rs

```diff
@@ -250,41 +258,99 @@ impl TraceCompiler {
             ));
         };
 
-        // FIXME implement support for arguments.
-        if !args.is_empty() {
-            return Err(CompileError::Unimplemented("call with args".to_owned()));
+        if args.len() > 6 {
+            return Err(CompileError::Unimplemented(
+                "call with spilled args".to_owned(),
+            ));
         }
 
-        // Generally speaking, we don't know what the callee will clobber, so we save our registers
-        // before doing the call. For now we assume that the caller is OK with RAX being clobbered
-        // though.
-        for reg in 1..X86_86_NUM_REGS {
+        // Figure out where the return value (if there is one) is going.
+        let dest_reg: Option<u8> = if let Some(d) = dest {
+            Some(self.local_to_reg(d.local)?)
+        } else {
+            None
+        };
+
+        // Save Sys-V callee save registers to the stack, but skip the one (if there is one) that
+        // will store the return value. It's safe to assume the caller expects this to be
+        // clobbered.
+        //
+        // FIXME: Note that we don't save rax. Although this is a callee save register, the way the
+        // tests currently work is they check the last value returned at the end of the trace. This
+        // value is assumed to remain in rax. If we were to restore rax, we'd break that. Note that
+        // the register allocator never gives out rax for this precise reason.
+        //
+        // rdi, rsi, rdx, rcx, r8, r9, r10, r11.
+        let save_regs = [7, 6, 2, 1, 8, 9, 10, 11]
```

Comment:
Maybe [remove_item](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.remove_item) is a bit shorter here and easier to read, though it is only in nightly, so I'll leave that decision up to you.

Reply:
Well, our ykrustc fork is technically nightly, so sounds good to me.

Reply:
Technically the channel you use is called `dev` :)

https://github.com/softdevteam/ykrustc/blob/1e8ac85b61ba402d6bff972b48933f94f0e3f893/config.toml.example#L349-L352

```rust
# The "channel" for the Rust build to produce. The stable/beta channels only
# allow using stable features, whereas the nightly and dev channels allow using
# nightly features
#channel = "dev"
```

Reply:
Did you want to do this too or leave as is?

Reply:
It's kind of annoying, because `dest_reg` is `Option<u8>` not `u8`. I'd have to wrap every element of the list in `Some`.